File: aib/app/services/extraction.py
"""
Intake data extraction — runs Claude on a completed transcript, stores results.
Mirrors server/src/services/extraction.js

Supports both PostgreSQL and in-memory fallback.
"""

from __future__ import annotations

import json
from datetime import datetime, timezone

from app.database import query, query_one, is_db_available
from app.services import session as session_svc
from app.services import anthropic_client
from app.prompts.extraction import EXTRACTION_PROMPT

# In-memory intake store (used when DB is unavailable)
_intakes: dict[str, dict] = {}


def extract_intake_data(session_id: str) -> dict:
    """
    Build a transcript from the session's messages, run extraction via Claude,
    and store the structured result.
    """
    messages = session_svc.get_messages(session_id)
    if not messages:
        raise ValueError("No messages found for session")

    # Build transcript including attachment metadata
    lines: list[str] = []
    for m in messages:
        line = f"{m['role'].upper()}: {m['content']}"
        attachments = m.get("attachments") or []
        if isinstance(attachments, str):
            attachments = json.loads(attachments)
        if attachments:
            file_list = ", ".join(
                f"{a['filename']} ({a['type']})" for a in attachments
            )
            line += f"\n[Attachments: {file_list}]"
        lines.append(line)

    transcript = "\n\n".join(lines)
    full_prompt = EXTRACTION_PROMPT + transcript

    extracted = anthropic_client.extract(full_prompt)
    intake = _store_intake(session_id, extracted)
    return intake


def _store_intake(session_id: str, data: dict) -> dict:
    """Upsert extracted data into the intakes table or in-memory store."""
    if is_db_available():
        return query_one(
            """INSERT INTO intakes (
                session_id, company_name, dba, address, fein,
                business_description, annual_revenue, employees_total,
                employees_ft_pt, annual_payroll, policy_type,
                total_limit_requested, existing_policies, financials_available,
                records_count, cyber_incidents, shareholders_5pct,
                fye_financials, last_12mo_revenue, epl_international_entities,
                epl_claims, erisa_plan_assets, media_content_type,
                contract_required, contract_provided, uploaded_documents,
                client_questions_flagged, additional_notes, raw_extraction
            ) VALUES (
                %s, %s, %s, %s, %s, %s, %s, %s, %s, %s,
                %s, %s, %s, %s, %s, %s, %s, %s, %s, %s,
                %s, %s, %s, %s, %s, %s, %s, %s, %s
            )
            ON CONFLICT (session_id) DO UPDATE SET
                company_name = EXCLUDED.company_name,
                dba = EXCLUDED.dba,
                address = EXCLUDED.address,
                fein = EXCLUDED.fein,
                business_description = EXCLUDED.business_description,
                annual_revenue = EXCLUDED.annual_revenue,
                employees_total = EXCLUDED.employees_total,
                employees_ft_pt = EXCLUDED.employees_ft_pt,
                annual_payroll = EXCLUDED.annual_payroll,
                policy_type = EXCLUDED.policy_type,
                total_limit_requested = EXCLUDED.total_limit_requested,
                existing_policies = EXCLUDED.existing_policies,
                financials_available = EXCLUDED.financials_available,
                records_count = EXCLUDED.records_count,
                cyber_incidents = EXCLUDED.cyber_incidents,
                shareholders_5pct = EXCLUDED.shareholders_5pct,
                fye_financials = EXCLUDED.fye_financials,
                last_12mo_revenue = EXCLUDED.last_12mo_revenue,
                epl_international_entities = EXCLUDED.epl_international_entities,
                epl_claims = EXCLUDED.epl_claims,
                erisa_plan_assets = EXCLUDED.erisa_plan_assets,
                media_content_type = EXCLUDED.media_content_type,
                contract_required = EXCLUDED.contract_required,
                contract_provided = EXCLUDED.contract_provided,
                uploaded_documents = EXCLUDED.uploaded_documents,
                client_questions_flagged = EXCLUDED.client_questions_flagged,
                additional_notes = EXCLUDED.additional_notes,
                raw_extraction = EXCLUDED.raw_extraction
            RETURNING *""",
            (
                session_id,
                data.get("company_name"),
                data.get("dba"),
                data.get("address"),
                data.get("fein"),
                data.get("business_description"),
                data.get("annual_revenue"),
                data.get("employees_total"),
                data.get("employees_ft_pt"),
                data.get("annual_payroll"),
                data.get("policy_type"),
                data.get("total_limit_requested"),
                data.get("existing_policies"),
                data.get("financials_available"),
                data.get("records_count"),
                data.get("cyber_incidents"),
                data.get("shareholders_5pct"),
                data.get("fye_financials"),
                data.get("last_12mo_revenue"),
                data.get("epl_international_entities"),
                data.get("epl_claims"),
                data.get("erisa_plan_assets"),
                data.get("media_content_type"),
                data.get("contract_required", False),
                data.get("contract_provided", False),
                json.dumps(data.get("uploaded_documents", [])),
                json.dumps(data.get("client_questions_flagged", [])),
                data.get("additional_notes"),
                json.dumps(data),
            ),
        )

    # In-memory fallback
    intake = {
        "session_id": session_id,
        "created_at": datetime.now(timezone.utc).isoformat(),
        **data,
    }
    _intakes[session_id] = intake
    return intake


def get_intake(session_id: str) -> dict | None:
    """Get intake data for a session."""
    if is_db_available():
        return query_one(
            "SELECT * FROM intakes WHERE session_id = %s",
            (session_id,),
        )
    return _intakes.get(session_id)


def list_intakes() -> list[dict]:
    """List all completed intakes."""
    if is_db_available():
        return query(
            """SELECT i.*, s.status AS session_status, s.created_at AS session_created_at
               FROM intakes i
               JOIN sessions s ON s.id = i.session_id
               ORDER BY i.created_at DESC"""
        )
    return list(_intakes.values())

File: aib/server/src/services/extraction.js
const db = require('../config/database');
const anthropic = require('./anthropic');
const EXTRACTION_PROMPT = require('../prompts/extraction');

async function extractIntakeData(sessionId) {
  const messagesResult = await db.query(
    `SELECT role, content, attachments FROM messages WHERE session_id = $1 ORDER BY created_at ASC`,
    [sessionId]
  );

  if (messagesResult.rows.length === 0) {
    throw new Error('No messages found for session');
  }

  const transcript = messagesResult.rows
    .map(m => {
      let line = `${m.role.toUpperCase()}: ${m.content}`;
      const attachments = typeof m.attachments === 'string'
        ? JSON.parse(m.attachments)
        : m.attachments;
      if (attachments && attachments.length > 0) {
        const fileList = attachments.map(a => `${a.filename} (${a.type})`).join(', ');
        line += `\n[Attachments: ${fileList}]`;
      }
      return line;
    })
    .join('\n\n');

  const fullPrompt = EXTRACTION_PROMPT + transcript;
  const extracted = await anthropic.extract(fullPrompt);
  const intake = await storeIntake(sessionId, extracted);
  return intake;
}

async function storeIntake(sessionId, data) {
  const result = await db.query(
    `INSERT INTO intakes (
      session_id, company_name, dba, address, fein,
      business_description, annual_revenue, employees_total,
      employees_ft_pt, annual_payroll, years_in_business,
      prior_carrier, retroactive_date, desired_effective_date,
      policy_type, total_limit_requested, existing_policies,
      financials_available, claims_history, records_count,
      cyber_incidents, cyber_mfa, cyber_backups, cyber_endpoint_security,
      shareholders_5pct, fye_financials, last_12mo_revenue,
      epl_international_entities, epl_claims, erisa_plan_assets,
      media_content_type, contract_required, contract_provided,
      uploaded_documents, client_questions_flagged, additional_notes, raw_extraction
    ) VALUES (
      $1, $2, $3, $4, $5, $6, $7, $8, $9, $10,
      $11, $12, $13, $14, $15, $16, $17, $18, $19, $20,
      $21, $22, $23, $24, $25, $26, $27, $28, $29, $30,
      $31, $32, $33, $34, $35, $36, $37
    )
    ON CONFLICT (session_id) DO UPDATE SET
      company_name = EXCLUDED.company_name,
      dba = EXCLUDED.dba,
      address = EXCLUDED.address,
      fein = EXCLUDED.fein,
      business_description = EXCLUDED.business_description,
      annual_revenue = EXCLUDED.annual_revenue,
      employees_total = EXCLUDED.employees_total,
      employees_ft_pt = EXCLUDED.employees_ft_pt,
      annual_payroll = EXCLUDED.annual_payroll,
      years_in_business = EXCLUDED.years_in_business,
      prior_carrier = EXCLUDED.prior_carrier,
      retroactive_date = EXCLUDED.retroactive_date,
      desired_effective_date = EXCLUDED.desired_effective_date,
      policy_type = EXCLUDED.policy_type,
      total_limit_requested = EXCLUDED.total_limit_requested,
      existing_policies = EXCLUDED.existing_policies,
      financials_available = EXCLUDED.financials_available,
      claims_history = EXCLUDED.claims_history,
      records_count = EXCLUDED.records_count,
      cyber_incidents = EXCLUDED.cyber_incidents,
      cyber_mfa = EXCLUDED.cyber_mfa,
      cyber_backups = EXCLUDED.cyber_backups,
      cyber_endpoint_security = EXCLUDED.cyber_endpoint_security,
      shareholders_5pct = EXCLUDED.shareholders_5pct,
      fye_financials = EXCLUDED.fye_financials,
      last_12mo_revenue = EXCLUDED.last_12mo_revenue,
      epl_international_entities = EXCLUDED.epl_international_entities,
      epl_claims = EXCLUDED.epl_claims,
      erisa_plan_assets = EXCLUDED.erisa_plan_assets,
      media_content_type = EXCLUDED.media_content_type,
      contract_required = EXCLUDED.contract_required,
      contract_provided = EXCLUDED.contract_provided,
      uploaded_documents = EXCLUDED.uploaded_documents,
      client_questions_flagged = EXCLUDED.client_questions_flagged,
      additional_notes = EXCLUDED.additional_notes,
      raw_extraction = EXCLUDED.raw_extraction
    RETURNING *`,
    [
      sessionId,
      data.company_name || null,
      data.dba || null,
      data.address || null,
      data.fein || null,
      data.business_description || null,
      data.annual_revenue || null,
      data.employees_total || null,
      data.employees_ft_pt || null,
      data.annual_payroll || null,
      data.years_in_business || null,
      data.prior_carrier || null,
      data.retroactive_date || null,
      data.desired_effective_date || null,
      data.policy_type || null,
      data.total_limit_requested || null,
      data.existing_policies || null,
      data.financials_available || null,
      data.claims_history || null,
      data.records_count || null,
      data.cyber_incidents || null,
      data.cyber_mfa ?? null,
      data.cyber_backups ?? null,
      data.cyber_endpoint_security ?? null,
      data.shareholders_5pct || null,
      data.fye_financials || null,
      data.last_12mo_revenue || null,
      data.epl_international_entities || null,
      data.epl_claims || null,
      data.erisa_plan_assets || null,
      data.media_content_type || null,
      data.contract_required || false,
      data.contract_provided || false,
      JSON.stringify(data.uploaded_documents || []),
      JSON.stringify(data.client_questions_flagged || []),
      data.additional_notes || null,
      JSON.stringify(data),
    ]
  );

  return result.rows[0];
}

async function getIntake(sessionId) {
  const result = await db.query(
    `SELECT * FROM intakes WHERE session_id = $1`,
    [sessionId]
  );
  return result.rows[0] || null;
}

async function listIntakes() {
  const result = await db.query(
    `SELECT i.*, s.status as session_status, s.created_at as session_created_at
     FROM intakes i
     JOIN sessions s ON s.id = i.session_id
     ORDER BY i.created_at DESC`
  );
  return result.rows;
}

module.exports = { extractIntakeData, getIntake, listIntakes };
